### ISSUE TITLE
Use custom thrift decoder to improve speed of parsing parquet metadata

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#5854]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Part of https://github.com/apache/arrow-rs/issues/5853

Parsing the parquet metadata takes substantial time and most of that time is spent in decoding the thrift format (@XiangpengHao is quantifying this in https://github.com/apache/arrow-rs/issues/5770)

**Describe the solution you'd like**
Improve the thrift decoder speed

**Describe alternatives you've considered**
@jhorstmann reports  on https://github.com/apache/arrow-rs/issues/5775 that he made a prototype of this:

              I had an attack of "not invented here" syndrome the last few days 😅 and worked on an alternative code generator for thrift, that would allow me to more easily try out some changes to the generated code. The repo can be found at <https://github.com/jhorstmann/compact-thrift/> and the output for `parquet.thrift` at <https://github.com/jhorstmann/compact-thrift/blob/main/src/main/rust/tests/parquet.rs>.

The current output is still doing allocations for string and binary, but running the benchmarks from <https://github.com/tustvold/arrow-rs/tree/thrift-bench> shows some nice improvements. This is the comparison with current arrow-rs code, so both versions should be doing the same amount of allocations:

```
decode metadata      time:   [32.592 ms 32.645 ms 32.702 ms]

decode metadata new  time:   [17.440 ms 17.476 ms 17.532 ms]
```

So incidentally very close to that 2x improvement.

The main difference in the code should be avoiding most of the abstractions from `TInputProtocol` and avoiding stack moves by directly writing into default-initialized structs instead of moving from local variables.

_Originally posted by @jhorstmann in https://github.com/apache/arrow-rs/issues/5775#issuecomment-2131307588_
            

**Additional context**
Sub Tasks
- [ ] https://github.com/apache/arrow-rs/issues/8537
- [ ] https://github.com/apache/arrow-rs/issues/5775

**Comment by jhorstmann** [comment]
Thanks @alamb for creating this tracking issue. I've slowly continued working on my code at [jhorstmann/compact-thrift](https://github.com/jhorstmann/compact-thrift/) and benchmarks are looking good. So good in fact that adapting it on top of #5777 the performance hotspot shifts to the conversion functions from generated thrift types to internal types.

I would love to get some feedback on the code, and whether there would be a preference to integrate the parquet definitions into the arrow-rs repo, or publish them separately.

The generated and runtime code is also structured in a way that it would not be too crazy to write bindings to custom types by hand.

Direct links to [the generated code](https://github.com/jhorstmann/compact-thrift/blob/main/src/main/rust/parquet/src/format.rs) and to [the runtime library](https://github.com/jhorstmann/compact-thrift/blob/main/src/main/rust/runtime/src/lib.rs).

**Comment by XiangpengHao** [comment]
FWIW, by simply moving this field to heap (i.e., `Option<Statistics>` -> `Option<Box<Statistics>>`), we can get 30% performance improvement (as will show in blog #5770).
https://github.com/apache/arrow-rs/blob/087f34b70e97ee85e1a54b3c45c5ed814f500b0a/parquet/src/format.rs#L3407

The `Option<Statistics>` occupies 136 bytes even if the file does not have stats at all (i.e., the field is `None`); this not only slows down decoding (due to poor memory locality) but also causes high memory consumption when decoding metadata (parquet-rs consumes 10MB memory per MB of metadata).


I think this example motivates custom parquet type definitions and, thus, custom thrift decoder.

**Comment by alamb** [comment]
> FWIW, by simply moving this field to heap (i.e., `Option<Statistics>` -> `Option<Box<Statistics>>`), we can get 30% performance improvement (as will show in blog #5770).

This code is in https://github.com/apache/arrow-rs/pull/5856 for anyone who is curious

**Comment by alamb** [comment]
> I would love to get some feedback on the code, and whether there would be a preference to integrate the parquet definitions into the arrow-rs repo, or publish them separately.

Hi @jhorstmann  -- I had a look at https://github.com/jhorstmann/compact-thrift/tree/main (very cool)

Some initial reactions:
 Writing a code generator in Kotlin is a neat idea, but I think it might make the barrier to contribution high (now people need to know Rust and Kotlin (and the associated toolchains, etc)

Also, I keep thinking if we are going to have a parquet-rs specific implementation,  why use a code generator at all? Maybe we could simply  hand code a decoder directly that uses the runtime library 

Given how infrequently the parquet spec changes, a hand rolled parser might be reasonable (though I will admin that the [generated format.rs](https://github.com/jhorstmann/compact-thrift/blob/main/src/main/rust/parquet/src/format.rs) is substantial 🤔 ). We can probably ensure compatibility with round trip testing of generated rust code 🤔

**Comment by jhorstmann** [comment]
> Writing a code generator in Kotlin is a neat idea, but I think it might make the barrier to contribution high (now people need to know Rust and Kotlin (and the associated toolchains, etc)

I agree, in the context of arrow-rs this is probably a bigger barrier to contribute than the existing C++ based thrift code generator.

Maybe the amount of code could be simplified and made easier to change by hand with the use of some macros. The most tricky part of the code generation, difficult to replicate in a macro, might be the decision of which structs require lifetime annotations.

**Comment by alamb** [comment]
The more I think about this the more I am convinced that the fastest thing to do would be to decode directly from thrift --> the parquet-rs structs. Perhaps we could follow the tape decoding model of the csv or json parsers in this repo 🤔 

Decoding to intermediate thrift structures which are then throw away seems like an obvious source of improvement

**Comment by jhorstmann** [comment]
It occurred to me that the thrift definitions consist entirely of valid rust tokens, and so should be parseable using declarative macros. The result of that experiment can be seen in #5909, the complete macro can be found at https://github.com/jhorstmann/compact-thrift/blob/main/src/main/rust/runtime/src/macros.rs

**Comment by alamb** [comment]
> It occurred to me that the thrift definitions consist entirely of valid rust tokens, and so should be parseable using declarative macros.

That is really (really) cool @jhorstmann 

Maybe we could even use the declarative macros to creating a parser that avoids intermediates, by providing callbacks rather than building structs 🤔

**Comment by etseidl** [comment]
Bumping this rather than creating a new issue. Also rolling in #7909 and #6129.

Here's what I'm planning:

1. Add more thrift processing benchmarks
2. Reduce use of `parquet::format` as much as possible, especially in publicly exposed data structures like `FileMetaData`.
3. Create a custom thrift parser to decode directly to the structures created in step 2. Part of this task will address #7909 by correctly dealing with unknown union values for `LogicalType` and `ColumnOrder`. This step will also leverage the macros developed by @jhorstmann (https://github.com/jhorstmann/compact-thrift).
4. Use parser from 3 internally to read non-exposed structures such as the page headers.
5. Add ability to write new structures directly to thrift-encoded bytes.
6. Remove the `format` module.
7. Explore opportunities for further speed ups. Examples include skipping row groups and projected columns, not decoding page statistics, halt processing after reading schema.

Hopfully I can have all of the above ready in time for 57.0.0 😅

**Comment by etseidl** [comment]
@alamb if it's ok with you, I plan to create a branch for this work after #8037 is merged. Since all of these changes will be breaking, it will be nice to have a place to make incremental progress that the community can keep tabs on and provide feedback. I'll of course keep the branch up to date with main to ease merging down the road.

**Comment by alamb** [comment]
> Bumping this rather than creating a new issue. Also rolling in [#7909](https://github.com/apache/arrow-rs/issues/7909) and [#6129](https://github.com/apache/arrow-rs/issues/6129).
> 
> Here's what I'm planning:
> 
> 1. Add more thrift processing benchmarks
> 2. Reduce use of `parquet::format` as much as possible, especially in publicly exposed data structures like `FileMetaData`.

This will be awesome -- I have likewise found the format module very hard to work with

> 4. Create a custom thrift parser to decode directly to the structures created in step 2. Part of this task will address [[Parquet] reader appears to have bug when supporting unknown sort orders #7909](https://github.com/apache/arrow-rs/issues/7909) by correctly dealing with unknown union values for `LogicalType` and `ColumnOrder`. This step will also leverage the macros developed by [@jhorstmann](https://github.com/jhorstmann) (https://github.com/jhorstmann/compact-thrift).

> 6. Use parser from 3 internally to read non-exposed structures such as the page headers.

This will be awesome -- when I was working on the parquet push decoder I also found that the existing thrift decoder copies data that is already in memory *AGAIN* using the `std::io::Read` trait -- it would be awesome if the new decoder took data in in memory `Buffer`. 

Note sure what your plans are, but I would like to recommend a "push" decoder that avoids having to do IO at all - similarly to what I have in https://github.com/apache/arrow-rs/pull/7997. A taste from `parquet/src/file/metadata/push_decoder.rs`

```rust
/// The `ParquetMetaDataPushDecoder` needs to know the file length.
/// let mut decoder = ParquetMetaDataPushDecoder::try_new(file_len).unwrap();
/// // try to decode the metadata. If more data is needed, the decoder will tell you what ranges
/// loop {
///     match decoder.try_decode() {
///        Ok(DecodeResult::Data(metadata)) => { return Ok(metadata); } // decode successful
///        Ok(DecodeResult::NeedsData(ranges)) => {
///           // The decoder needs more data
///           //
///           // In this example, we call a function that returns the bytes for each given range.
///           // In a real application, you would likely read the data from a file or network.
///           let data = ranges.iter().map(|range| get_range(range)).collect();
///           // Push the data into the decoder and try to decode again on the next iteration.
///           decoder.push_ranges(ranges, data).unwrap();
///        }
///        Ok(DecodeResult::Finished) => { unreachable!("returned metadata in previous match arm") }
///        Err(e) => return Err(e),
///     }
/// }
/// # }
```

> 8. Add ability to write new structures directly to thrift-encoded bytes.
> 9. Remove the `format` module
> 11. Explore opportunities for further speed ups. Examples include skipping row groups and projected columns, not decoding page statistics, halt processing after reading schema.> > Hopfully I can have all of the above ready in time for 57.0.0 😅

Sounds ambitious! 

However, I am around to review things -- l please feel free to ping me and I'll try my best to get them reviewed

**Comment by etseidl** [comment]
Thanks @alamb. I will be leaning quite heavily on you and others to get the design of all this right. The push decoder looks interesting...the same principles applied to the metadata could avoid unnecessary re-reads of data when the prefetch size is too small. I'll keep that in mind moving forward.

**Comment by alamb** [comment]
> Thanks [@alamb](https://github.com/alamb). I will be leaning quite heavily on you and others to get the design of all this right. The push decoder looks interesting...the same principles applied to the metadata could avoid unnecessary re-reads of data when the prefetch size is too small. I'll keep that in mind moving forward.

Thank you -- I hope to have a PR up with (just) a Push based MetadataDecoder design for your review -- I think it may make implementing custom thrift decoding simpler

**Comment by etseidl** [comment]
Update: yesterday I finished the end-to-end remodel and removed all dependencies on the `format` module (except 2 in `bin` and `benches`). After #8237 I have 6 more PRs stacked and ready to merge. Then a *lot* of polishing and fixing documentation, as well as documenting how the macros and all the write support work.

As variant and geo work progresses I'll have to merge that in somehow too.

One thing I've already worked in is skipping reading of page header statistics unless they are specifically requested. That's one small example of how writing our own decoder enables cutting down on needless decoding work.

**Comment by alamb** [comment]
Filed a ticket to test with DataFusion
 - https://github.com/apache/arrow-rs/issues/8513

I think @etseidl  is pretty close to getting the thrift-remodel branch ready to merge, see here for more details
- https://github.com/apache/arrow-rs/pull/8476#issuecomment-3348930333

**Comment by alamb** [comment]
@etseidl  how far away are we from being able to merge the gh thrift-remodel branch into main?

That seems like a large enough change that the longer the bake time we have (so early adopters can test, find bugs, etc) the better 🤔

**Comment by etseidl** [comment]
I think we can merge now (but I'm biased). No need to wait on #8528, that can merge to main instead.

My only hesitation is there will only be a few weeks to test, unless we delay 57 into November.

**Comment by alamb** [comment]
> I think we can merge now (but I'm biased). No need to wait on [#8528](https://github.com/apache/arrow-rs/pull/8528), that can merge to main instead.
> 
> My only hesitation is there will only be a few weeks to test, unless we delay 57 into November.

How about a mid way ground -- I'll test the thrift-remodel branch with DataFusion and if that looks ok, we can merge it to main. If we hit issues we can consider delaying some more

**Comment by etseidl** [comment]
Sounds good. I'm going to put up a draft PR from the remodel branch so we can get more eyes on it. Once you finish your testing (and I mine), we can mark it ready for review.

**Comment by alamb** [comment]
I also started working on a blog about this feature: 
- https://github.com/apache/arrow-rs/issues/8537

**Comment by adriangb** [comment]
Omg it got merged!!


![](https://i.imgflip.com/a8hl5u.jpg)
